kv,rpc,server: consolidate RPC clients creation

cthumuluru-crdb opened the pull request:
To support DRPC and reuse the connection management logic in [Connection](https://github.com/cockroachdb/cockroach/blob/master/pkg/rpc/connection.go#L54) and [Peer](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/peer.go#L121), they were refactored to use generics.

Previously, most code established connections to peers using [nodedialer](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/nodedialer/nodedialer.go#L37-L41) or [rpc.Context](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/context.go#L2004-L2016), subsequently creating an RPC client. Supporting DRPC would require updating numerous call-sites to conditionally dial `gRPC` or `DRPC` connections, introducing significant boilerplate code.

This commit consolidates RPC client creation logic in the `kvserver` package. Call sites were updated to dial RPC clients than dialing connections. Subsequent work will ensure the appropriate connection type is dialed and the corresponding client is created. RPC clients will be unified to use a consistent interface for both `gRPC` and `DRPC` connections.

Epic: [CRDB-48923](https://cockroachlabs.atlassian.net/browse/CRDB-48923)
Informs: https://github.com/cockroachdb/cockroach/issues/147757
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147606)
<!-- Reviewable:end -->

tbg reviewed (commented):
The pattern isn't too pretty, but it works and I can't immediately come up with something better.

One thought I had was that it would be nicer if the helpers were defined "next to" the generated client code for each RPC call, but this likely runs into dependency issues because the connection class would import the `rpc` package from the proto packages. Still, this seems solvable:

First, move `rpc.ConnectionClass` into a subpkg `rpcbase.ConnClass`.

Then, say, in `ctpb`, in a `.go` file near `SideTransportClient`'s, you could write something like this:

```go

type nodeDialer interface { // or move this to `rpcbase.NodeDialerI` to reuse everywhere, probably better
  DialDRPC(context.Context, roachpb.NodeID, rpcbase.ConnClass) ...
  DialGRPC(context.Context, roachpb.NodeID, rpcbase.ConnClass) ...
}

func DialSideTransportClient(nd nodeDialer, nodeID roachpb.NodeID, class rpcbase.ConnectionClass) ctpb.SideTransportClient {
  if !drpc {
    conn, err := nd.Dial(ctx, nodeID, class)
    // ...
  }
  // ...
}
```

If you can figure out the details, that might be the superior option because it keeps the extra code close to each client interface. I think it would also reduce the size of each change.

shubhamdhama reviewed (commented):
LGTM! It made my life easy for server registration-related changes. I've put some minor comments. I'll approve once they are resolved.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-:-ORjxCYmCvvqVyLOenU0:b-ayvlcp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb and @shaikzakiriitm)

___
*[`pkg/kv/kvserver/closedts/sidetransport/rpc_clients.go` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjuE5jBim69eHNZb-y:-ORjuE5jBim69eHNZb-z:bgyx7wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/closedts/sidetransport/rpc_clients.go#L11)):*
> ```Go
> )
> 
> type nodeClientDialer struct {
> ```

I think you'd want to update this one too.
___
*[`pkg/kv/kvserver/rpc_clients.go` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjtJXU-gtMPDQ-qXb2:-ORjtJXV05SBP-VSw3pk:b-yrh8fc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/rpc_clients.go#L14)):*
> ```Go
> type nodeClientDialer nodedialer.Dialer
> 
> func (d *nodeClientDialer) DialMultiRaftClient(
> ```

`nodeClientDialer.NewMultiRaftClient` is more idiomatic IMO. The intention of "Dial" is implied because it's a method on a Dialer. What do you think?



_Code quote:_
```Go
DialMultiRaftClient
```
___
*[`pkg/kv/kvserver/storeliveness/rpc_clients.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjw0GEDiHEBEvovn4N:-ORjw0GEDiHEBEvovn4O:bshbsyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/storeliveness/rpc_clients.go#L35)):*
> ```Go
> 	nd := (*nodedialer.Dialer)(d)
> 	return nd.GetCircuitBreaker(nodeID, class)
> }
> ```

Having these functions that are unrelated to RPC clients here feels a bit awkward to me. What do you think about keeping the existing `nodedialer.Dialer` in `Transport.dialer` and then, wherever we need the client, we can directly do this:
```go
		client, err := AsClientDialer(t.dialer).DialStoreLivenessClient(ctx, toNodeID, connClass)
```

Or, put a method `nodeClientDialer.AsNodeDialer`, which returns back the `(*nodedialer.Dialer)(d)`



_Code quote:_
```Go
func (d *nodeClientDialer) GetCircuitBreaker(
	nodeID roachpb.NodeID, class rpc.ConnectionClass,
) (*circuit2.Breaker, bool) {
	nd := (*nodedialer.Dialer)(d)
	return nd.GetCircuitBreaker(nodeID, class)
}
```
___
*[`pkg/kv/kvserver/storeliveness/rpc_clients.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjvjGTCXg5tuoEPRzB:-ORjvjGTCXg5tuoEPRzC:bx9ljix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/storeliveness/rpc_clients.go#L42)):*
> ```Go
> 	nd := (*nodedialer.Dialer)(d)
> 	return nd.Dial(ctx, nodeID, class)
> }
> ```

This function is not used I guess.



_Code quote:_
```Go
func (d *nodeClientDialer) Dial(
	ctx context.Context, nodeID roachpb.NodeID, class rpc.ConnectionClass,
) (_ *grpc.ClientConn, err error) {
	nd := (*nodedialer.Dialer)(d)
	return nd.Dial(ctx, nodeID, class)
}
```


<!-- Sent from Reviewable.io -->

cthumuluru-crdb committed:
kv,rpc,server: consolidate RPC clients creation

To support DRPC and reuse the connection management logic in
[Connection](https://github.com/cockroachdb/cockroach/blob/master/pkg/rpc/connection.go#L54)
and [Peer](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/peer.go#L121),
they were refactored to use generics.

Previously, most code established connections to peers using
[nodedialer](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/nodedialer/nodedialer.go#L37-L41)
or [rpc.Context](https://github.com/cockroachdb/cockroach/blob/8fed6e8481da1373020ab2cc95d222df02f61721/pkg/rpc/context.go#L2004-L2016),
subsequently creating an RPC client. Supporting DRPC would require
updating numerous call-sites to conditionally dial `gRPC` or `DRPC`
connections, introducing significant boilerplate code.

This commit consolidates RPC client creation logic in the `kvserver`
package. Call sites were updated to dial RPC clients than dialing
connections. Subsequent work will ensure the appropriate connection type
is dialed and the corresponding client is created. RPC clients will be
unified to use a consistent interface for both `gRPC` and `DRPC`
connections.

Epic: CRDB-48923
Informs: https://github.com/cockroachdb/cockroach/issues/147757
Release note: none

shaikzakiriitm commented on pkg/kv/kvserver/raft_transport.go:
```diff
@@ -844,13 +844,10 @@ func (t *RaftTransport) startProcessNewQueue(
 			t.connectionMu.Unlock()
 		}()
-		conn, err := t.dialer.Dial(ctx, toNodeID, class)
+		client, err := t.dialer.DialMultiRaftClient(ctx, toNodeID, class)
```
Isn't this model necessitating us to create a wrapper function (with probable prefix `Dial`) for each of the client apis? Looks like more boiler plate work, or will these functions be auto-generated?

cthumuluru-crdb replied:
Done.  Please refer to the other comments.
<!-- Reviewable comment -ORovkhlB5GeBXJq3nZW:bsmzk1x -->
<!-- Sent from Reviewable.io -->

shaikzakiriitm commented on pkg/kv/kvserver/rpc_clients.go:
```diff
@@ -0,0 +1,65 @@
+
+type nodeClientDialer nodedialer.Dialer
+
+func (d *nodeClientDialer) DialMultiRaftClient(
```
We need to have a guideline documented (someplace accessible) that as long as grpc and drpc co-exist, devs need to follow this strategy pattern to choose the appropriate client depending upon probably a feature flag.

cthumuluru-crdb replied:
Yeah. We will document it and share it with broader database team. 
<!-- Reviewable comment -ORovIrh2sKZgH07Y957:b-91qdk2 -->
<!-- Sent from Reviewable.io -->

shaikzakiriitm commented on pkg/kv/kvserver/storeliveness/rpc_clients.go:
```diff
@@ -0,0 +1,46 @@
+
+func (d *nodeClientDialer) DialStoreLivenessClient(
+	ctx context.Context, nodeID roachpb.NodeID, class rpc.ConnectionClass,
+) (slpb.StoreLivenessClient, error) {
```
Can you please share an example of how the signature looks like with a drpc client here? Afaik, drpc client interface type looks like `DRPCStoreLivenessClient`, which is a different client type.

cthumuluru-crdb replied:
Done. We have a different PR that unifies the RPC client interfaces across gRPC and DRPC. One example is available here: https://github.com/cockroachdb/drpc/pull/4#issuecomment-2929254397. 
<!-- Reviewable comment -ORouCTyBvDFTYslIjts:b-3likzy -->
<!-- Sent from Reviewable.io -->

shaikzakiriitm reviewed (commented):
Overall, changes look good. I have left a few minor questions. Thanks!

Can you please explicitly add a line or two on the advantages of the chosen approach over the others called out in the pr description, for future reference. Probably call out how the chosen approach is better in terms of lesser boilerplate code or more concise in terms of usage at the places of client calls.

cthumuluru-crdb reviewed (commented):
@tbg, I tried to place the helpers right next to the generate protobuf changes but then I ran into cyclic dependencies. I followed you suggestion and moved `rpc.ConnectionClass` and introduced an interface for `NodeDialer`. Now I'm able to resolve the cyclic dependencies and placed the helpers closer to the generated code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-:-ORosXhp3GBF1tdgkOTM:bhp27d)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shaikzakiriitm, @shubhamdhama, and @tbg)



<!-- Sent from Reviewable.io -->

cthumuluru-crdb commented on pkg/kv/kvserver/rpc_clients.go:
```diff
@@ -0,0 +1,65 @@
+	ctx context.Context, nodeID roachpb.NodeID, class rpc.ConnectionClass,
+) (MultiRaftClient, error) {
+	nd := (*nodedialer.Dialer)(d)
+	if true { // TODO(server) gRPC
```
Done!

cthumuluru-crdb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147606)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shaikzakiriitm, @shubhamdhama, and @tbg)

___
*[`pkg/kv/kvserver/rpc_clients.go` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjtJXU-gtMPDQ-qXb2:-ORovT4K5l7HMI39Gy8O:b-oq6bgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/rpc_clients.go#L14)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

`nodeClientDialer.NewMultiRaftClient` is more idiomatic IMO. The intention of "Dial" is implied because it's a method on a Dialer. What do you think?
</blockquote></details>

Yeah. Your suggestion makes sense for a method on a dialer, it makes sense. Since the helper functions also dial a connection, I prefer keeping the dial part for them.
___
*[`pkg/kv/kvserver/closedts/sidetransport/rpc_clients.go` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjuE5jBim69eHNZb-y:-ORovGbi8yKnhS2Rrltu:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/closedts/sidetransport/rpc_clients.go#L11)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

I think you'd want to update this one too.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/storeliveness/rpc_clients.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjw0GEDiHEBEvovn4N:-ORou9ZV5Ph0UaF5FrqV:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/storeliveness/rpc_clients.go#L35)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

Having these functions that are unrelated to RPC clients here feels a bit awkward to me. What do you think about keeping the existing `nodedialer.Dialer` in `Transport.dialer` and then, wherever we need the client, we can directly do this:
```go
		client, err := AsClientDialer(t.dialer).DialStoreLivenessClient(ctx, toNodeID, connClass)
```

Or, put a method `nodeClientDialer.AsNodeDialer`, which returns back the `(*nodedialer.Dialer)(d)`
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/storeliveness/rpc_clients.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORjvjGTCXg5tuoEPRzB:-ORou27E67qmcBB6JHuf:berx43o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3ebe8347e07416197bba0762a4f8e21f5641a42/pkg/kv/kvserver/storeliveness/rpc_clients.go#L42)):*
<details><summary><i>Previously, shubhamdhama (Shubham Dhama) wrote…</i></summary><blockquote>

This function is not used I guess.
</blockquote></details>

Done. We are not using dialer anymore. I changed the code to use helpers.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 93 of 93 files at r2, 15 of 15 files at r4, 17 of 17 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147606)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shaikzakiriitm and @shubhamdhama)

___
*[`-- commits` line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORuCCwF9bbG29vM73zy:-ORuCCwF9bbG29vM73zz:b-uf0aqf):*
Surely you have an epic here?
___
*[`pkg/rpc/rpcbase/nodedialer.go` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORuCnb89l1_fc2mNaLF:-ORuCnb89l1_fc2mNaLG:b8vzwyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad644e89267ca08fdaa55ed8e5c66b7b0b7129b2/pkg/rpc/rpcbase/nodedialer.go#L15)):*
> ```Go
> )
> 
> const TODODRPC = true
> ```

It's a bit odd that now the use of this is

```go
if TODORPC {
  // do the thing that is NOT drpc
}
// unimplemented DRPC-enabled path
```

Intuitively I would expect the bool to be false, i.e. the it should read

```
if !TODORPC {
  // do the thing that is NOT drpc
}
// unimplemented DRPC-enabled path
```

Might be worth flipping this, since it's likely to stay for a while. Your call.


<!-- Sent from Reviewable.io -->

cthumuluru-crdb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147606)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shaikzakiriitm, @shubhamdhama, and @tbg)

___
*[`-- commits` line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORuCCwF9bbG29vM73zy:-ORukXhX6NLSpxRGCLDt:b-4wdja8):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Surely you have an epic here?
</blockquote></details>

Yes. I'll update it before committing the changes.
___
*[`pkg/rpc/rpcbase/nodedialer.go` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147606#-ORuCnb89l1_fc2mNaLF:-ORukbDb3Z-xlAHa5ab4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad644e89267ca08fdaa55ed8e5c66b7b0b7129b2/pkg/rpc/rpcbase/nodedialer.go#L15)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It's a bit odd that now the use of this is

```go
if TODORPC {
  // do the thing that is NOT drpc
}
// unimplemented DRPC-enabled path
```

Intuitively I would expect the bool to be false, i.e. the it should read

```
if !TODORPC {
  // do the thing that is NOT drpc
}
// unimplemented DRPC-enabled path
```

Might be worth flipping this, since it's likely to stay for a while. Your call.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->
